lms/djangoapps/instructor_analytics/basic.py

```diff
@@ -316,24 +303,33 @@ def student_submissions(course, all_students=False):
                 unit_name = unit.display_name_with_default
 
-                row = [section_name.encode('utf-8'), subsection_name.encode('utf-8'), unit_name.encode('utf-8'), '']
+                parent_metadata = [section_name.encode('utf-8'), subsection_name.encode('utf-8'), unit_name.encode('utf-8')]
                 for component in unit.get_children():
                     if component.category == 'problem':
-                        row[-1] = component.display_name_with_default.encode('utf-8')
-                        modules = StudentModule.objects.filter(course_id=course.id, grade__isnull=False, module_state_key=component.location)
-                        if modules:
-                            answers = [''] * len(students)
-                            for module in modules:
-                                try:
-                                    state_dict = json.loads(module.state) if module.state else {}
-                                    raw_answers = state_dict.get("student_answers", {})
-                                except ValueError:
-                                    log.error("Student submissions: Could not parse module state for " +
-                                              "StudentModule id={}, course={}".format(module.id, course.id))
-                                    continue
-                                pretty_answers = u', '.join(u"{problem}={answer}".format(problem=problem, answer=answer) for (problem, answer) in raw_answers.items())
-                                answers[student_cols[module.student.username]] = pretty_answers.encode('utf-8')
-                            if any(answers):
-                                datarows.append(row + answers)
-
-    return header, datarows
+                        yield component, parent_metadata
+
+
+def student_submissions(course):
+    """
+    Yields student submissions for all problems in course for writing out to a CSV file.
+    """
+    order = 1
+    for problem, parent_metadata in iterate_problem_components(course):
+        problem_info = parent_metadata + [problem.display_name_with_default.encode('utf-8'), order, problem.location]
+        modules = StudentModule.objects.filter(course_id=course.id, grade__isnull=False, module_state_key=problem.location).order_by('student__username')
+        if modules:
+            has_answer = False
+            for module in modules:
+                try:
+                    state_dict = json.loads(module.state) if module.state else {}
+                    raw_answers = state_dict.get("student_answers", {})
+                except ValueError:
+                    log.error("Student submissions: Could not parse module state for " +
+                              "StudentModule id={}, course={}".format(module.id, course.id))
+                    continue
+                pretty_answers = u', '.join(u"{problem}={answer}".format(problem=problem, answer=answer) for (problem, answer) in raw_answers.items())
```

Comment:
I changed the outer one to be more clearer distinct, let me know if this is better